src/lib/storage/simdCAI_segment/simdCAI_segment_iterable.hpp

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include <algorithm>
+
+#include "include/codecfactory.h"
+#include "include/intersection.h"
+#include "include/frameofreference.h"
+
+#include "storage/segment_iterables.hpp"
+#include "storage/simdCAI_segment.hpp"
+
+#include "utils/performance_warning.hpp"
+
+namespace opossum {
+
+template <typename T>
+class SIMDCAISegmentIterable : public PointAccessibleSegmentIterable<SIMDCAISegmentIterable<T>> {
+ public:
+  using ValueType = T;
+
+  explicit SIMDCAISegmentIterable(const SIMDCAISegment<T>& segment) : _segment{segment} {}
+
+  template <typename Functor>
+  void _on_with_iterators(const Functor& functor) const {
+    _segment.access_counter[SegmentAccessCounter::AccessType::Sequential] += _segment.size();
+    auto begin = Iterator{_segment.encoded_values(), &_segment.null_values(), _segment.codec_id(), _segment.size(), ChunkOffset{0}};
+    auto end = Iterator{_segment.encoded_values(), &_segment.null_values(), _segment.codec_id(), _segment.size(), static_cast<ChunkOffset>(_segment.size())};
+
+    functor(begin, end);
+  }
+
+  template <typename Functor, typename PosListType>
+  void _on_with_iterators(const std::shared_ptr<PosListType>& position_filter, const Functor& functor) const {
+    _segment.access_counter[SegmentAccessCounter::access_type(*position_filter)] += position_filter->size();
+
+    using PosListIteratorType = decltype(position_filter->cbegin());
+    auto begin =
+        PointAccessIterator<PosListIteratorType>{_segment.encoded_values(), &_segment.null_values(), _segment.codec_id(), _segment.size(),
+                                                 position_filter->cbegin(), position_filter->cbegin()};
+    auto end =
+        PointAccessIterator<PosListIteratorType>{_segment.encoded_values(), &_segment.null_values(), _segment.codec_id(), _segment.size(),
+                                                 position_filter->cbegin(), position_filter->cend()};
+    functor(begin, end);
+  }
+
+  size_t _on_size() const { return _segment.size(); }
+
+ private:
+  const SIMDCAISegment<T>& _segment;
+
+ private:
+  class Iterator : public AbstractSegmentIterator<Iterator, SegmentPosition<T>> {
+    public:
+      using ValueType = T;
+      using IterableType = SIMDCAISegmentIterable<T>;
+      using EndPositionIterator = typename pmr_vector<ChunkOffset>::const_iterator;
+
+    public:
+      explicit Iterator(const std::shared_ptr<const pmr_vector<uint32_t>>& encoded_values,
```

Comment:
I wouldn't pass a shared_ptr here but rather a pointer to the actual encoded values.